src/main/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperation.java

```diff
@@ -0,0 +1,125 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package hudson.plugins.build_timeout.operations;
+
+import org.kohsuke.stapler.DataBoundConstructor;
+import hudson.Extension;
+import hudson.model.AbstractBuild;
+import hudson.model.Cause;
+import hudson.model.BuildListener;
+import hudson.model.ParametersAction;
+import hudson.model.Run;
+import hudson.plugins.build_timeout.BuildTimeOutOperation;
+import hudson.plugins.build_timeout.BuildTimeOutOperationDescriptor;
+
+/**
+ * Abort the build.
+ */
+public class AbortAndRestartOperation extends BuildTimeOutOperation {
+    private final int maxRestarts;
+    
+    /**
+     * @return max restarts.
+     */
+    public int getMaxRestarts() {
+        return maxRestarts;
+    }
+    
+    @DataBoundConstructor
+    public AbortAndRestartOperation(int maxRestarts){
+        this.maxRestarts = maxRestarts;
+    }
+       
+    /**
+     * @param build
+     * @param listener
+     * @param effectiveTimeout
+     * @return
+     * @see hudson.plugins.build_timeout.BuildTimeOutOperation#perform(hudson.model.AbstractBuild, hudson.model.BuildListener, long)
+     */
+    @Override
+    public boolean perform(AbstractBuild<?, ?> build, BuildListener listener, long effectiveTimeout) {
+        
+        new AbortOperation().perform(build, listener, effectiveTimeout);
+                
+        if (checkMaxRestarts(build)) {
+
+            ParametersAction action = build.getAction(ParametersAction.class);
+            build.getRootBuild().getProject().scheduleBuild(0, new BuildTimeoutAbortAndRestartCause(build), action);
+        }
+        return true;
+    }
+    
+    public class BuildTimeoutAbortAndRestartCause extends Cause {
+        
+
+        Run<?, ?> build;
+        
+        /**
+         * Constructor.
+         * 
+         * @param s
+         *            The reason/cause for restart.
+         */
+        public BuildTimeoutAbortAndRestartCause(Run<?, ?> up) {
+            super();
+            this.build = up;
+        }
+        
+        @Override
+        public String getShortDescription() {
+            return "Build Timeout - Abort and Restart: Aborted by build no.: " + build.getNumber();
+        
+        }
+    }
+    
+    private boolean checkMaxRestarts(AbstractBuild<?, ?> build) {
+        if (this.maxRestarts <= 0) {
+            return true;
+        }
+        int count = 0;
+
+        // count the number of restarts for the current project
+        while (build != null) {
+            
+            if(build.getCause(BuildTimeoutAbortAndRestartCause.class) != null){
+                count++;
+            }
+         
+            if (count >= this.maxRestarts) {
+                return false;
+            }
+            build = build.getPreviousBuild();
+        }

```

Comment:
I believe this is not what you want to do.
For example, if maxRestarts is 2:
* build1: aborted. triggers rtey.
* build2: retry of build1, aborted. triggers retry.
* build3: retry of build2, aborted. exceeded retrying.

In this case, new builds never trigger retries as there's already two retying builds in the history.
Using `BuildTimeoutAbortAndrestartCause#build` is recommended.